_posts/2016-05-22-th1eadin0g1.markdown

```diff
@@ -0,0 +1,202 @@
+
+### NSOperation
+
+An NSThread is a pretty simple model for doing threading. It allows you to run one function on a background thread, but what if you needed something more complicated, e.g. scheduling, priorities, state or management. This is the space that `NSOperation`s fill.
+
+There are three main ways people use `NSOperation`s: 
+
+* **Using a custom subclass** - happens when you have specific work to do, you've may have used these yourself via [AFNetworking](https://cocoapods.org/pods/AFNetworking) which used to create [NSOperation subclasses](http://cocoadocs.org/docsets/AFNetworking/2.6.3/Classes/AFHTTPRequestOperation.html) to do it's networking.
+
+   A lot of our most critical code in Folio, the sync engine, is a collection of NSOperation subclasses. I did a video [explaining it all this year](http://artsy.github.io/blog/2016/02/12/Code-Review-Energy-Sync/).
+
+   [Yuliya](https://github.com/Yuliya-Kaleda), a friend of mine,let me know that the `NSOperation` abstraction is similar to Android's [AsyncTask](https://developer.android.com/reference/android/os/AsyncTask.html) - which has a _really_ interesting pattern of wrapping off the main thread work, and then doing some execution on the main thread.
+
+* **Using `NSInvocationOperation`** - This is actually what I recommended in my blog post in 2012, it's essentially a simple wrapper around `NSTask` which calls a specific method.
+* **Using `NSBlockOperation`** - This is a nice option when you want to run some code from a block. More on blocks later. 
+
+`NSOperation`s on their own are  pretty simple objects, but they allow you to define dependency graphs. So you could have one operation taking the JPG from the file system, then another operation that takes the image data and converts it in an `NSImage`, then the latter can depend on the former. Now you have some heirarchy. That's nice.
+
+No discussion about `NSOperation` would be complete without `NSOperationQueue`. This is the abstraction you use to limit how many operations can be running at the same time. Say you had a lot of images to download, you could use an operation queue with a limit of three concurrent operations. They also provide an easy way to determine priority too, so you could have a higher priority queue for images you would see, then a lower priority queue for images that are not on the screen at the moment.
+
+This is only a small overview, for more information, I strongly recommend reading [NSHipster's NSOperation](http://nshipster.com/nsoperation/) post. 
+
+### Grand Central Dispatch (GCD)
+
+This is the most used tool in modern Objective-C threading. GCD relies on blocks. As I know some android people are reading this, a block is a way of treating code in similar way that you would variables. These can be passed between methods, a very common case is callbacks. They let you do some work, then let something know that you've completed the work. The advantage being that you can keep all the code in one place.
```

Comment:
Maybe mention a closure, which is a more common concept that other devs might be familiar with.